lambda/switch.ml

```diff
@@ -659,9 +662,14 @@ let rec pkey chan  = function
           and right = {s with cases=right} in
 
           if i=1 && (lim+ctx.off)=1 && get_low cases 0+ctx.off=0 then
-            make_if_ne
-              ctx.arg 0
-              (c_test ctx right) (c_test ctx left)
+            if lcases = 2 && get_high cases 1+ctx.off = 1 then
```

Comment:
I don't know that part very well, but I wonder whether this also handles
```
type a =
  | A
  | B
  | C

let f = function
  | A -> 4
  | B
  | C -> 5
```

Reply:
It intentionally does not. The semantics of Lambda aren't documented, but my understanding is that `Lifthenelse` expects its argument to be either `0` or `1` (rather than a C-style zero-or-nonzero test). So, this optimisation only applies when the argument is known to be one of those two values.

Your example is compiled as before, because the case-split in `f` is into two cases, one for `0` and one for the interval `[1,2]`. Thus `lcases = 2`, but `get_high cases 1+ctx_off = 2`.

Reply:
I think that the Lifthenelse is C-style. That's the reason why match on None/Some is compiled as a single if.

```OCaml
type a =
  | A
  | C of int

let f = function
  | A -> 4
  | C _ -> 5
```

```
  (let (f/1005 = (function param/1006 (if param/1006 5 4)))
    (makeblock 0 f/1005)))
```
You can use it (and some 'magic') if you want to test the semantics of Lifthenelse

By the way, I realize that this could also benefit.
```OCaml
type a =
  | A
  | B of int
  | C of int

let f = function
  | A -> 4
  | B _
  | C _ -> 5
```

Reply:
Ah, I see! I hadn't noticed that the compiler already used `Lifthenelse` on values which were not booleans. That means I can use a less conservative check.

With this more liberal optimisation, the change is much simpler: it just replaces `if (x != 0)` with the equivalent `if (x)`.

This generates optimised Lambda for your first example. Your second remains the same, as the pre-Switch code generates a `Pisint` primitive to distinguish the block from integer constructors. (I don't think that there's any particular benefit from changing this).